lambda-buffers-codegen/src/LambdaBuffers/Codegen/Rust/Print/LamVal.hs

```diff
@@ -0,0 +1,397 @@
+ method on the target type:
+ ```rs
+ <TargetType>::to_plutus_data(...)
+ ```
+ In case of `jsonConstructor`, we print the target type in a turbofish syntax:
+ ```rs
+ lbr_prelude::json::json_constructr::<TargetType>::toPlutusData(...)
+ ```
+-}
+printRefE :: MonadPrint m => LV.Ref -> m (Doc ann)
+printRefE ref = do
+  qvn <- LV.resolveRef ref
+  case ref of
+    (argTy : _, builtin)
+      | builtin == "toPlutusData"
+          || builtin == "fromPlutusData"
+          || builtin == "toJson"
+          || builtin == "fromJson" -> do
+          lamTyDoc <- printLamTy argTy
+          methodDoc <- R.printRsValName . R.qualifiedEntity <$> LV.importValue qvn
+          return $ angles lamTyDoc <> R.doubleColon <> methodDoc
+      | builtin == "jsonConstructor" -> do
+          lamTyDoc <- printLamTy argTy
+          methodDoc <- R.printRsQValName <$> LV.importValue qvn
+          return $ methodDoc <> R.doubleColon <> langle <> lamTyDoc <> rangle
+    _ -> R.printRsQValName <$> LV.importValue qvn
+
+printLamTy :: MonadPrint m => LT.Ty -> m (Doc ann)
+printLamTy (LT.TyRef tyRef) = return $ R.printTyRef tyRef
+printLamTy (LT.TyVar tyVar) = return $ R.printTyVar tyVar
+printLamTy (LT.TyApp ty gens _) = do
+  tyDoc <- printLamTy ty
+  gensDoc <- for gens printLamTy
+  return $ tyDoc <> R.encloseGenerics gensDoc
+printLamTy _ = throwInternalError "Unexpected LamTy expression."
+
+printValueE :: MonadPrint m => PC.TyDefs -> LV.ValueE -> m (Doc ann)
+printValueE _ (LV.VarE v) = return $ pretty v
+printValueE _ (LV.RefE ref) = printRefE ref
+printValueE iTyDefs (LV.LamE lamVal) = printLamE iTyDefs lamVal
+printValueE iTyDefs (LV.AppE funVal argVal) = printAppE iTyDefs funVal argVal
+printValueE iTyDefs (LV.CaseE sumTy caseVal ctorCont) = printCaseE iTyDefs sumTy caseVal ctorCont
+printValueE iTyDefs (LV.CtorE qctor prodVals) = printCtorE iTyDefs qctor prodVals
+printValueE iTyDefs (LV.RecordE qrec vals) = printRecordE iTyDefs qrec vals
+printValueE iTyDefs (LV.FieldE fieldName recVal) = printFieldE iTyDefs fieldName recVal
+printValueE iTyDefs (LV.ProductE qprod vals) = printProductE iTyDefs qprod vals
+printValueE iTyDefs (LV.LetE prodTy prodVal letCont) = printLetE iTyDefs prodTy prodVal letCont
+printValueE _ (LV.IntE i) = return $ pretty i
+printValueE iTyDefs (LV.CaseIntE intVal cases otherCase) = printCaseIntE iTyDefs intVal cases otherCase
+printValueE iTyDefs (LV.ListE vals) = printNewListE iTyDefs vals
+printValueE iTyDefs (LV.CaseListE listVal cases otherCase) = printCaseListE iTyDefs listVal cases otherCase
+printValueE _ (LV.TextE txt) = printTextE txt
+printValueE iTyDefs (LV.CaseTextE txtVal cases otherCase) = printCaseTextE iTyDefs txtVal cases otherCase
+printValueE iTyDefs (LV.TupleE l r) = printTupleE iTyDefs l r
+printValueE _ (LV.ErrorE err) = throwInternalError $ "LamVal error builtin was called " <> err
+
+-- | This is a hack, to help Rust figure out types of closures (lambda expressions)
```

Comment:
nit; HACK(szg251): Needed to help Rust figure out types of closures (lambda expressions).